Libraries/LibPthread/pthread.cpp

```diff
@@ -157,70 +167,233 @@ int pthread_sigmask(int how, const sigset_t* set, sigset_t* old_set)
     return 0;
 }
 
-int pthread_mutex_init(pthread_mutex_t* mutex, const pthread_mutexattr_t* attributes)
+#define PTHREAD_FUTEX_UNLOCKED 0u
+#define PTHREAD_FUTEX_LOCKED 1u
+#define PTHREAD_FUTEX_LOCKED_WITH_WAITERS 2u
+
+static int futex_lock(uint32_t& futex_addr, uint32_t futex_flags, const timespec* abstime = nullptr, bool no_wait = false)
 {
-    mutex->lock = 0;
-    mutex->owner = 0;
-    mutex->level = 0;
-    mutex->type = attributes ? attributes->type : PTHREAD_MUTEX_NORMAL;
-    return 0;
+    ASSERT(!(futex_flags & ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)));
+    for (;;) {
+        uint32_t expected = PTHREAD_FUTEX_UNLOCKED;
+        if (AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED, AK::MemoryOrder::memory_order_acquire)) {
+            atomic_thread_fence(AK::MemoryOrder::memory_order_release);
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": acquired @ " << &futex_addr;
+#endif
+            return 0;
+        }
+        if (no_wait)
+            return EBUSY;
+        if (expected == PTHREAD_FUTEX_LOCKED_WITH_WAITERS || AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED_WITH_WAITERS, AK::MemoryOrder::memory_order_acquire)) {
```

Comment:
ditto weak variant

Reply:
Also I don't think you need acquire here: if the cmp_xcg fails, you're continuing to the next iteration, where you already use acquire.

Reply:
Using `memory_order_relaxed` in a loop is not a good idea, either. Doing so can supposedly (some architectures) cause a extreme load on the memory bus, affecting performance much more than e.g. `memory_order_acquire`.

Reply:
Really??

I'm not an expert by any means, but I can't imagine why relaxed could be slower then acquire, since it seems that acquire does everything relax does and then some. So it'd always be correct (if inefficient) to replace any relaxed atomic op by an acquire op — so how can it be any slower?

Reply:
Well the acquire would be slower but not saturate the memory bus as much as relaxed. It's not an issue on x86 but supposedly on some others (maybe even arm?). I can't remember where I read this, but it was essentially the exact same question: why one would not just use relaxed and when successful, do an acquire memory fence and bail the loop.

Reply:
Here's the quote I was referring to:

> This "optimization" is however a) useless and b) may in fact hurt: a) Since the thread will be busily spinning on a blocked spinlock, it does not matter if it will waste the CPU cycles with just "exchange" operations or with both useless "exchange" and "acquire" operations. b) A tight "exchange" loop without any memory-synchronizing instruction introduced through an "acquire" operation will on some systems monopolize the memory subsystem and degrade the performance of other system components.

As per the "Discussion" section of the [Spinlock example](https://www.boost.org/doc/libs/1_72_0/doc/html/atomic/usage_examples.html#boost_atomic.usage_examples.example_spinlock).